shared_model/utils/variant_deserializer.hpp

```diff
@@ -62,15 +62,20 @@ namespace shared_model {
          * @param v result variant
          */
         template <class V, class T = typename V::types, class Archive>
-        static V invoke(const Archive &ar, int which) {
+        static V invoke(Archive &&ar, int which) {
           if (which == 0) {
             using head_type = typename boost::mpl::front<S>::type;
-            return V(head_type(ar));
+            using variant_head_type = typename boost::mpl::front<T>::type;
+            static_assert(
```

Comment:
The motivation of this code is not so obvious. Maybe write a comment about purposes of using this assert?